runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,243 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local api = vim.api
+local M = {}
+
+-- [[
+-- TODO:
+-- [ ] only request for visible range
+-- ]]
+--
+---@class lsp.inlay_hint.bufstate
+---@field version integer
+---@field client_hint table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
+
+---@type table<integer, lsp.inlay_hint.bufstate>
+local hint_cache_by_buf = setmetatable({}, {
+  __index = function(t, b)
+    local key = b > 0 and b or api.nvim_get_current_buf()
+    return rawget(t, key)
+  end,
+})
+
+--- |lsp-handler| for the method `textDocument/inlayHint`
+--- Store hints for a specific buffer and client
+--- Resolves unresolved hints
+---
+function M.on_inlayhint(err, result, ctx, _)
+  if err then
+    local _ = log.error() and log.error('inlayhint', err)
+    return
+  end
+  local bufnr = ctx.bufnr
```

Comment:
Right here, you could return if `util.buf_version[bufnr] ~= bufnr`. This can really help performance when the buffer changes while the response is being calculated by the server for a request from a previous version. 

The decoration provider doesn't use hints if the version doesn't match anyway, so why process the response? This is part of what I was talking about in the other comment.

In the semantic tokens module, I tracked the most recent "active" request's version, but here we've added the associated version directly to the context. When a response arrives, the semantic token module only processes it if it's still valid and I think the same thing would work here.